feat(new source): New internal_logs source

bruceg opened the pull request:
Note that the logs observed by this source are limited by the `$LOGS` option at startup. Working around this limitation is difficult and potentially a significant performance impact, as it requires turning on all events up front. I'd appreciate if anybody has better ideas for dealing with this.

Additional known limitations for future work:
1. This is just the simple source with no filters. It forwards on all logs and traces generated internally. It would be useful for it to have options for a minimum level and/or for including and excluding certain types of logs.
2. This source does nothing at present about skipping logs resulting from transforms and sinks that are downstream from itself. @lukesteensen and I discussed it and figured on traversing the topology to build an exclusion list. However, this will be quite a bit of work to do compared to just getting this in place, so it would probably best be done as a follow-up issue.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Closes #3458

Commit:
Drop unused `bus` crate

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for broadcasting log events in the tracing lib

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add internal_logs source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix docs format

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
More metadata fixes

Signed-off-by: Bruce Guenter <bruce@timber.io>

MOZGIII commented on src/trace.rs:
```diff
@@ -39,3 +53,137 @@ pub fn init(color: bool, json: bool, levels: &str) {
+    #[inline]
+    fn event(&self, event: &tracing::Event<'_>) {
+        if self.sender.receiver_count() > 0 {
+            let _ = self.sender.send(event.into()); // Ignore errors
```
This might drop the events if the `sender` channel is full, right? It the desired behaviour imo, just asking if it does indeed do that.

bruceg replied:
Yes, this is a "best effort" kind of send. This is effectively required behavior, because the alternative (blocking if the channel is full) could quite trivially cause vector to lock up. Reporting the error could similarly lead to an infinite loop.

However, while investigating this behavior, I see that *receiving* messages when the channel is full can produce a temporary error, which should not cause the `internal_logs` source to stop (which it will now). I'll have to fix that.

MOZGIII commented on src/trace.rs:
```diff
@@ -39,3 +53,137 @@ pub fn init(color: bool, json: bool, levels: &str) {
 }
+
+pub fn subscribe() -> Option<Receiver<Event>> {
+    SENDER.get().map(|sender| sender.subscribe())
```
Cool, didn't know this API existed!

MOZGIII commented on src/trace.rs:
```diff
@@ -39,3 +53,137 @@ pub fn init(color: bool, json: bool, levels: &str) {
+        maker.0.insert("metadata".into(), data.into());
+        maker.0.insert("timestamp".into(), now.into());
+
+        let log: LogEvent = maker.0.into_iter().collect();
```
It might be more efficient to build `LogEvent`s here directly.

bruceg replied:
There's probably a better way. I'll investigate.

MOZGIII reviewed (approved):
Looks good, but from prior experience, I'm getting an idea it would be good to know the overhead of this upfront. How about adding a couple of benches with/without the internal logs source so we can compare the perf?

bruceg commented:
> It would be good to know the overhead of this upfront. How about adding a couple of benches with/without the internal logs source so we can compare the perf?

I agree it would be good to measure, but I don't expect a big problem. I'll explain.

There are actually four different scenarios:
1. Without this PR in the source (baseline)
2. With this PR in the source but the `sources-internal_logs` feature not enabled.
3. With `sources-internal_logs` enabled but no `internal_logs` source configured.
4. With both of the above enabled.

I would expect all of 1-3 to have the same performance (for other features, obviously), within the noise. The `BroadcastSubscriber` wrapper is purely inline, and so is a no-op for performance, except for `fn event`. This function has a guard on it that counts the number of subscribers before doing any of its work. That guard is (should be) an unlocked read and branch, which should have a 100% branch prediction rate, making it too a no-op except for case 4.

Case #4 is the interesting one, but this is where the user has explicitly opted in to take the performance hit. We would expect vector to slow down for that case. Of course, we do want to minimize that slow down, but it's a much smaller concern than cases 1-3 above.

All of this, of course, assumes logging is actually in the hot path. If we're not generating more than a dozen or so messages per second, it won't be, and then performance doesn't matter as much.

Commit:
Merge remote-tracking branch 'origin/master' into internal-logs-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Create LogEvent directly, avoids an extra trip through BTreeMap

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Ignore the "lagged" error in the receive loop

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix testing issue with setting the logs broadcast multiple times

This has the unfortunate side effect of turning on error logs for all
tests, which make them considerably more noisy.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into internal-logs-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

lukesteensen reviewed (approved):
Nice! This looks like a great start. I agree we should address the things you mentioned and carefully measure performance before we enable the feature by default, but this will be good to start experimenting with.

bruceg commented:
I think this is ready to merge, but the metadata doesn't check out and I couldn't figure out why: https://github.com/timberio/vector/pull/5346/checks?check_run_id=1501970891#step:9:423

Does anybody else have some insight into what I've done wrong here? @binarylogic?

binarylogic committed:
Fix docs

Signed-off-by: binarylogic <bjohnson@binarylogic.com>

binarylogic commented:
@bruce I fixed the docs in the above commits. I'm unsure what to put for examples, could you update those with accurate examples? Thanks.

bruceg commented:
Thanks @binarylogic. I will update the examples, yes.

Commit:
Update examples for the `module_path` and `target` metadata

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix up error message

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into internal-logs-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

leebenson commented:
Cross-posting -- this PR appears to have removed metrics from the API / `vector top` described in #5511

API unit tests failed.

bruceg commented:
Fixed in #5529, I hope
